docs/adr/adr-003-march2022-testnet.md

```diff
@@ -0,0 +1,174 @@
+work otherwise, leading to last-minute solutions, like having to hand both the celestia **light** and **full** node a 
+“trusted” hash of a header from the already-running chain so that it can sync from that point and start listening for 
+new headers.
+
+#### Proposed new architecture: [`BlockService` is only responsible for reconstructing the block from Shares handed to it by the `ShareService`](https://github.com/celestiaorg/celestia-node/issues/251).
+Right now, the `BlockService` is in charge of fetching new blocks from the core node, erasure coding them, generating 
+DAH, generating `ExtendedHeader`, broadcasting `ExtendedHeader` to `HeaderSub` network, and storing the block data 
+(after some validation checks).
+
+Instead, a **full** node will rely on `ShareService` sampling to fetch us *enough* shares to reconstruct the block 
+inside of `BlockService`. Contrastingly, a **bridge** node will not do block reconstruction via sampling, but rather 
+rely on the `header.CoreSubscriber` implementation of `header.Subscriber` for blocks. `header.CoreSubscriber` will 
+handle listening for new block events from the core node via RPC, erasure code the new block, generate the 
+`ExtendedHeader` and pipe the erasure coded block through to `BlockService` via an internal subscription.
+
+### `HeaderSync` optimizations
+* Implement disconnect toleration 
+
+### Unbonding period handling
+The **light** node is prone to the highly theoretical but still possible long-range attack. To mitigate it, we should 
+introduce an additional `trustPeriod` variable (equal to unbonding period) which applies to headers. Suppose a node 
+starts with the period between subjective head and objective head being higher than the unbonding period - 
+in that case, the **light** node must not trust the subjective head anymore, specifically its `ValidatorSet`. Therefore, 
+instead of syncing subsequent headers on top of the untrusted subjective head, the Node should request a new objective 
+head from the `trustedPeer` and set it as a new trusted subjective head.
+
+<hr style="border:1px solid gray"> </hr>
+
+## Nice to have
+
+### `ShareService` optimizations
+* Implement parallelization for retrieving shares by namespace. This
+  [issue](https://github.com/celestiaorg/celestia-node/issues/184) is already being worked on.
+* NMT/Shares/Namespace storage optimizations:
+  * Right now we prepend to each Share 17 additional bytes, Luckily, for each reason why the prepended bytes were added, 
+  there is an alternative solution: It is possible to get NMT Node type indirectly, without serializing the type itself
+  by looking at the amount of links. To recover the namespace of the erasured data, we should not encode namespaces into 
+  the data itself. It is possible to get the namespace for each share encoded in inner non-leaf nodes of the NMT tree. 
+* Pruning for shares. 
```

Comment:
I think another rather important nice to have not mentioned here, is that syncing past headers should request headers not only from one peer, but again following the tendermint light client security model, double-check with a randomly chosen ["witness"](https://github.com/tendermint/tendermint/blob/02d456b8b8274088e8d3c6e1714263a47ffe13ac/light/client.go#L138-L141) here as well.
See: https://github.com/tendermint/tendermint/blob/02d456b8b8274088e8d3c6e1714263a47ffe13ac/light/client.go#L154-L161

> The light client maintains a simple address book containing addresses of full nodes that it can pick as primary and secondaries. To obtain a new light block, the light client first does verification with the primary, and then cross-checks the light block (and the trace of light blocks that led to it) with the secondaries using this specification.

https://github.com/tendermint/spec/blob/master/spec/light-client/detection/detection_003_reviewed.md#context-of-this-document